### PR TITLE
Feature/statistics on homepage #31

### PR DESCRIPTION
**Pull request by Christopher-Sandbach**
The layout needs a small amendment, ie statsBox and statsToggle. All other files will remain the same

### CODE REVIEW

**Comment by Christopher-Sandbach** [comment]
envrc.template will need updating

File: src/UI/state/home/actions/getHomepageStats.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+const date = new Date();
+
+export const getHomepageStats = createAsyncThunk(
+  'home/getHomepageStats',
+  async (_, thunkAPI) => {
+    const response = await fetchGraphQlData(
+      getHomepageAnalytics(0, date.getTime(), 'hour', 'Europe%2FLondon')
+    );
+    const data = response.data.getHomepageAnalytics;
+    thunkAPI.dispatch(updateStats(data));
```
Comment (andykitch) [inline src/UI/state/home/actions/getHomepageStats.ts:15]: What happens here if `getHomepageAnalytics` fails? The analytics server might be down for some reason, which would probably cause this call to fail. Does the UI then fall over as well? We probably just want it to not display the stats if this fails.

File: src/UI/components/home/statsBox/statsBox.tsx [hunk 2]
```diff
@@ -64,41 +71,59 @@ export default function StatsBox() {
             width: 'fit-content',
           }}
         >
-          <AnalyticsIcon sx={isMobile ? statsIconMobile : statsIconBrowser} />
-          <Typography color="primary" variant="h4" sx={{ fontSize: '5vw' }}>
+          {/* <AnalyticsIcon sx={isMobile ? statsIconMobile : statsIconBrowser} /> */}
+          {/* <Typography color="primary" variant="h4" sx={{ fontSize: '5vw' }}>
             Statistics
-          </Typography>
+          </Typography> */}
```
Comment (andykitch) [inline src/UI/components/home/statsBox/statsBox.tsx:77]: Commented out code?

File: src/UI/components/home/statsBox/statsBox.test.tsx [hunk 3]
```diff
@@ -12,10 +12,8 @@ describe('StatsBox component', () => {
...
   it('renders the currently hardcoded stats', () => {
     render(<StatsBox />, state);
-    expect(screen.getByText('Downloads')).toBeInTheDocument();
-    expect(screen.getByText('Data Points')).toBeInTheDocument();
-    expect(screen.getByText('Countries')).toBeInTheDocument();
-    expect(screen.getByText('Citations')).toBeInTheDocument();
-    expect(screen.getByText('Species')).toBeInTheDocument();
+    expect(screen.getByText('vector records')).toBeInTheDocument();
+    expect(screen.getByText('visited...')).toBeInTheDocument();
+    expect(screen.getByText('by..')).toBeInTheDocument();
```
Comment (andykitch) [inline src/UI/components/home/statsBox/statsBox.test.tsx:17]: It'd be nice to have a test for the initial action dispatch, in case someone removes it.

File: src/Docker/docker-compose.yml [hunk 4]
```diff
@@ -87,7 +87,12 @@ services:
       DATABASE_TYPE: postgresql
       BASE_PATH: /analytics
       CLIENT_IP_HEADER : "x-original-ip"
+      ANALYTICS_ADMIN_PASSWORD: $VA_ANAL_ADMIN_PASSWORD
+      ANALYTICS_API_URL: 'http://localhost:3003/analytics/api'
+      NEXT_PUBLIC_ANALYTICS_ID: 'fa6b40f4-9ba3-4f80-936c-2696c34f62db'
```
Comment (andykitch) [inline src/Docker/docker-compose.yml:92]: I'm not sure these are in the right place here - should they be in the API bit?

File: src/Docker/docker-compose.test-local.yml [hunk 5]
```diff
@@ -121,6 +121,9 @@ services:
       DATABASE_URL: postgresql://postgres:postgrespass@db:5432/umami
       DATABASE_TYPE: postgresql
       BASE_PATH: /analytics
+      ANALYTICS_ADMIN_PASSWORD: 'umami'
+      ANALYTICS_API_URL: 'http://localhost:3003/analytics/api'
+      NEXT_PUBLIC_ANALYTICS_ID: 'fa6b40f4-9ba3-4f80-936c-2696c34f62db'
```
Comment (andykitch) [inline src/Docker/docker-compose.test-local.yml:126]: Same comment as other docker compose file

File: src/API/src/analytics/analytics.service.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      throw e;
+    }
+  }
+
+  async metricsAnalytics(startAt: number, endAt: number, type: string) {
+    try {
+      const metrics = await lastValueFrom(
+        this.http
+          .get(
+            `${process.env.ANALYTICS_API_URL}/websites/${process.env.NEXT_PUBLIC_ANALYTICS_ID}/metrics?start_at=${startAt}&end_at=${endAt}&type=${type}`,
```
Comment (andykitch) [inline src/API/src/analytics/analytics.service.ts:97]: Most of this URL (up to `/metrics`) can be extracted out somewhere - you use it a few times. That'd get rid of the line length issue too.

Comment (Christopher-Sandbach) [reply src/API/src/analytics/analytics.service.ts]: I tried extracting it into a urlId property but it would return undefined and the associated tests would fail. I tried a const outside the scope of the class also and that failed similarly

**Comment by Christopher-Sandbach** [comment]
I'm unsure if the env variables will need to be set on the deployed server

File: src/API/src/analytics/analytics.service.ts (hunk 6 above)
Comment (andykitch) [reply src/API/src/analytics/analytics.service.ts:97]: Just stick a `const analyticsUrl = `${process.env.ANALYTICS_API_URL}/websites/${process.env.NEXT_PUBLIC_ANALYTICS_ID}`` at the top of the file, and use it where it's needed - does that work?

File: src/API/src/analytics/analytics.service.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      const approvedRecords = await this.occurrenceRepository
+        .createQueryBuilder('occurrence')
+        .leftJoin('occurrence.dataset', 'dataset')
+        .where('dataset.status = :status', { status: 'Approved' })
+        .getRawMany();
+      return approvedRecords.length;
+    } catch (e) {
+      this.logger.error(e);
+      throw e;
+    }
```
Comment (andykitch) [inline src/API/src/analytics/analytics.service.ts:125]: Have we removed the bit where we total up the download_count?

File: src/UI/state/home/actions/getHomepageStats.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  async (_, thunkAPI) => {
+    try {
+      const response = await fetchGraphQlData(
+        getHomepageAnalytics(0, date.getTime(), 'hour', 'Europe%2FLondon')
+      );
+      const data = response.data.getHomepageAnalytics;
+      thunkAPI.dispatch(serverValidation(true));
+      thunkAPI.dispatch(updateStats(data));
+    } catch (e: any) {
+      console.log(e.message);
```
Comment (andykitch) [inline src/UI/state/home/actions/getHomepageStats.ts:19]: Console log should be removed...

Comment (Christopher-Sandbach) [reply src/UI/state/home/actions/getHomepageStats.ts:19]: Don't we want to inform technical people if something has failed?

File: src/Docker/docker-compose.yml [hunk 9]
```diff
@@ -39,6 +39,9 @@ services:
       - AZURE_STORAGE_CONNECTION_STRING=$AZURE_STORAGE_CONNECTION_STRING
       - EMAIL_PASSWORD=$EMAIL_PASSWORD
       - ADMIN_EMAIL=vectoratlas@icipe.org
+      - ANALYTICS_ADMIN_PASSWORD=$VA_ANAL_ADMIN_PASSWORD
+      - ANALYTICS_API_URL=http://localhost:3003/analytics/api
+      - NEXT_PUBLIC_ANALYTICS_ID=fa6b40f4-9ba3-4f80-936c-2696c34f62db
```
Comment (andykitch) [inline src/Docker/docker-compose.yml:44]: I think this is probably different for the server

File: src/API/src/analytics/analytics.service.ts (hunk 7 above)
Comment (Christopher-Sandbach) [reply src/API/src/analytics/analytics.service.ts:125]: Yeah. When speaking with Ray he thought that may have caused confusion. So, I removed all the areas related to counting up the downloads. Now it's just number of vectors and individual download events (regardless of the number of rows downloaded)

File: src/Docker/docker-compose.yml (hunk 9 above)
Comment (Christopher-Sandbach) [reply src/Docker/docker-compose.yml:44]: > I think this is probably different for the server

This is what I'm wondering. I'm unsure how to setup the env variables so it is compatible with server

**Comment by andykitch** [comment]
> I'm unsure if the env variables will need to be set on the deployed server

Yeah, they will. There's advice for secret env variables at the bottom of the maintenance bit of the SMG

**Comment by Christopher-Sandbach** [comment]
image

File: src/UI/state/home/actions/getHomepageStats.ts (hunk 8 above)
Comment (andykitch) [reply src/UI/state/home/actions/getHomepageStats.ts:19]: Use the logger (in utils) - this does admittedly just console log it, but we'll swap it for something 'proper' later

File: src/Docker/docker-compose.yml (hunk 9 above)
Comment (andykitch) [reply src/Docker/docker-compose.yml:44]: Is this something we need to keep secret? If so, we'll need to follow the usual secret process. If not we can just stick the actual value in here. You can find the actual value by using DevTools on chrome, so I don't think it's that secret (it's `2b3b6da0-3e75-4d9e-be86-d588496a174e`)

File: .github/workflows/deployment.yml [hunk 10]
```diff
@@ -25,6 +25,8 @@ jobs:
           export VA_DB_URL="${{ secrets.TEST_SERVER_DB_URL }}"
           export AZURE_STORAGE_CONNECTION_STRING="${{ secrets.TEST_AZURE_STORAGE_CONNECTION_STRING }}"
           export EMAIL_PASSWORD="${{ secrets.TEST_EMAIL_PASSWORD }}"
+          export ANALYTICS_API_URL="${{ secrets.TEST_ANALYTICS_ADMIN_PASSWORD }}"
+          export ANALYTICS_ADMIN_PASSWORD="${{ secrets.TEST_ANALYTICS_API_URL }}"
```
Comment (andykitch) [inline .github/workflows/deployment.yml:29]: These are swapped around I think

File: src/API/src/analytics/analytics.service.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        map((res: any) => {
+          return res.data;
+        }),
+      ),
+  );
+  const website = websiteList.find((website) => {
+    return website.name === 'vector-atlas-test';
+  });
+  const uuid = website.websiteUuid;
+  return uuid;
```
Comment (andykitch) [inline src/API/src/analytics/analytics.service.ts:40]: Is this not the same as `NEXT_PUBLIC_ANALYTICS_ID`?

File: src/Docker/docker-compose.yml [hunk 12]
```diff
@@ -88,6 +91,8 @@ services:
       BASE_PATH: /analytics
       CLIENT_IP_HEADER : "x-original-ip"
     restart: always
+    ports:
+      - '3003:3000'
```
Comment (andykitch) [inline src/Docker/docker-compose.yml:95]: Is this change needed? I'm always nervous about unnecessary changes in this file :-)

File: src/UI/state/home/actions/getHomepageStats.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+export const getHomepageStats = createAsyncThunk(
+  'home/getHomepageStats',
+  async (_, thunkAPI) => {
+    try {
+      const response = await fetchGraphQlData(
+        getHomepageAnalytics(0, date.getTime(), 'hour', 'Europe%2FLondon')
+      );
+      const data = response.data.getHomepageAnalytics;
+      thunkAPI.dispatch(serverValidation(true));
```
Comment (andykitch) [inline src/UI/state/home/actions/getHomepageStats.ts:16]: Is this similar to a `statsLoading` flag? If so, can we change it to be a loading flag instead? It'd be more consistent with the rest of the actions.

File: src/Docker/docker-compose.yml (hunk 9 above)
Comment (Christopher-Sandbach) [reply src/Docker/docker-compose.yml:44]: The problem is that the ID changes. So the ID I put in will not be the same for the deployed instance. So, in the analytics service I have it as a property that is obtained from an API call - however, I have just noticed that we also use it in _app.tsx

So not sure how want to handle that 

```
function MyApp({ Component, pageProps }: AppProps) {
  useEffect(() => {
    store.dispatch(getFeatureFlags());
    store.dispatch(getMapStyles());
    store.dispatch(getUiVersion());
    store.dispatch(getApiVersion());
    store.dispatch(getTileServerOverlays());
  }, []);
  return (
    <>
      <Script
        async
        defer
        data-website-id={process.env.NEXT_PUBLIC_ANALYTICS_ID}
        src={process.env.NEXT_PUBLIC_ANALYTICS_URL}
      />
      <Provider store={store}>
        <ThemeProvider theme={theme}>
```

File: src/API/src/analytics/analytics.service.ts (hunk 11 above)
Comment (Christopher-Sandbach) [reply src/API/src/analytics/analytics.service.ts:40]: It is but I realized we can't hardcode it as an env variable because my id locally will not be the same when a umami instance is spun up when deployed

Comment (Christopher-Sandbach) [reply src/API/src/analytics/analytics.service.ts:40]: Unless that's something we can go in change manually but I am unfamiliar with deployments as a whole

File: src/Docker/docker-compose.yml (hunk 9 above)
Comment (andykitch) [reply src/Docker/docker-compose.yml:44]: If you put in the id for the deployed instance it'll be the same - that's the one I copied above. It should persist, as it's in the db. Seeing as we have it in the env variables for the UI, I think we can just stick it in for the API too.

File: src/API/src/analytics/analytics.service.ts (hunk 11 above)
Comment (andykitch) [reply src/API/src/analytics/analytics.service.ts:40]: See above :-)

**Review by andykitch — APPROVED** [review]

File: .github/workflows/deployment.yml [hunk 14]
```diff
@@ -25,6 +25,8 @@ jobs:
           export VA_DB_URL="${{ secrets.TEST_SERVER_DB_URL }}"
           export AZURE_STORAGE_CONNECTION_STRING="${{ secrets.TEST_AZURE_STORAGE_CONNECTION_STRING }}"
           export EMAIL_PASSWORD="${{ secrets.TEST_EMAIL_PASSWORD }}"
+          export ANALYTICS_API_URL="${{ secrets.TEST_ANALYTICS_API_URL }}" 
+          export ANALYTICS_ADMIN_PASSWORD="${{ secrets.TEST_ANALYTICS_ADMIN_PASSWORD }}"
```
Comment (andykitch) [inline .github/workflows/deployment.yml:29]: Have these been added to the github secrets?
